# Association Rules: Cosmetic Recommendation

## Exploration

In [21]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
from mlxtend.preprocessing import TransactionEncoder

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [22]:
dataset = pd.read_csv('Cosmetics.csv')
dataset.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Bag  Blush  Nail Polish  Brushes  Concealer  Eyebrow Pencils  Bronzer  \
0    0      1            1        1          1                0        1   
1    0      0            1        0          1                0        1   
2    0      1            0        0          1                1        1   
3    0      0            1        1          1                0        1   
4    0      1            0        0          1                0        1   

   Lip liner  Mascara  Eye shadow  Foundation  Lip Gloss  Lipstick  Eyeliner  
0          1        1           0           0          0         0         1  
1          1        0           0           1          1         0         0  
2          1        1           1           1          1         1         0  
3          0        0           0           1          0         0         1  
4          1        1           1           0          1         1         0

## Pre-processing

In [23]:
dataset = dataset.drop(['Trans. '], axis=1)
dataset = dataset.astype(bool)
dataset.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Bag  Blush  Nail Polish  Brushes  Concealer  Eyebrow Pencils  Bronzer  \
0  False   True         True     True       True            False     True   
1  False  False         True    False       True            False     True   
2  False   True        False    False       True             True     True   
3  False  False         True     True       True            False     True   
4  False   True        False    False       True            False     True   

   Lip liner  Mascara  Eye shadow  Foundation  Lip Gloss  Lipstick  Eyeliner  
0       True     True       False       False      False     False      True  
1       True    False       False        True       True     False     False  
2       True     True        True        True       True      True     False  
3      False    False       False        True      False     False      True  
4       True     True        True       False       True      True     False

## Model construction using association rules

### Generate all the frequent itemsets with support = 20%

In [24]:
frequent_itemsets=apriori(dataset,min_support=0.2,use_colnames=True)
frequent_itemsets

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support                  itemsets
0     0.363                   (Blush)
1     0.280             (Nail Polish)
2     0.442               (Concealer)
3     0.279                 (Bronzer)
4     0.234               (Lip liner)
5     0.357                 (Mascara)
6     0.381              (Eye shadow)
7     0.536              (Foundation)
8     0.490               (Lip Gloss)
9     0.322                (Lipstick)
10    0.457                (Eyeliner)
11    0.220        (Concealer, Blush)
12    0.204      (Mascara, Concealer)
13    0.201   (Concealer, Eye shadow)
14    0.231   (Concealer, Foundation)
15    0.218    (Concealer, Lip Gloss)
16    0.297     (Concealer, Eyeliner)
17    0.321     (Mascara, Eye shadow)
18    0.211  (Foundation, Eye shadow)
19    0.201   (Lip Gloss, Eye shadow)
20    0.356   (Foundation, Lip Gloss)
21    0.238    (Eyeliner, Foundation)
22    0.227     (Eyeliner, Lip Gloss)

### total number of frequent itemsets

In [27]:
len(frequent_itemsets)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


23

### Generate all the association rules with minimum support = 20% and minimum confidence = 70%

In [28]:
my_rules = association_rules(frequent_itemsets,metric="confidence",min_threshold=0.7)
my_rules

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents   consequents  antecedent support  consequent support  \
0     (Mascara)  (Eye shadow)               0.357               0.381   
1  (Eye shadow)     (Mascara)               0.381               0.357   
2   (Lip Gloss)  (Foundation)               0.490               0.536   

   support  confidence      lift  leverage  conviction  zhangs_metric  
0    0.321    0.899160  2.359999  0.184983    6.138417       0.896222  
1    0.321    0.842520  2.359999  0.184983    4.083050       0.930971  
2    0.356    0.726531  1.355468  0.093360    1.696716       0.514210

## Model deployment

### Show rule with highest confidence

In [33]:
my_rules.sort_values('confidence',ascending=False)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents   consequents  antecedent support  consequent support  \
0     (Mascara)  (Eye shadow)               0.357               0.381   
1  (Eye shadow)     (Mascara)               0.381               0.357   
2   (Lip Gloss)  (Foundation)               0.490               0.536   

   support  confidence      lift  leverage  conviction  zhangs_metric  
0    0.321    0.899160  2.359999  0.184983    6.138417       0.896222  
1    0.321    0.842520  2.359999  0.184983    4.083050       0.930971  
2    0.356    0.726531  1.355468  0.093360    1.696716       0.514210

### Explain the meaning of this rule

Transaction ที่มี Mascara มีโอกาส 89.9 % ที่จะมี Eye Shadow อยู่ด้วย

Transaction ที่มี Eye Shadow มีโอกาส 84.25 % ที่จะมี Mascara อยู่ด้วย

Transaction ที่มี Lip Gloss มีโอกาส 72.65 % ที่จะมี Foundation อยู่ด้วย

### Explain how support, confidence and lift are calculated?

support = freq(A,B) / N

confidence = support / (antecedent support)

lift = confident / (consequent support)

### Show example of using these rules to increase sales.

1. สร้างสินค้าที่มี mascara และ eye shadow ในราคาพิเศษ
2. วางสินค้าที่ confidence ไว้ใกล้ๆ กัน
3. ทำโปรโมชั่นแถม